docs/proposal-0005-jetstream-tooling-versioning.md

```diff
@@ -0,0 +1,62 @@
+
+## Keeping track of library and framework versions
+
+Jetstream is based on mozanalysis, metric-config-parser and some internal logic. When a new version of jetstream is released, a new Docker container gets pushed to [GCR](https://console.cloud.google.com/gcr/images/moz-fx-data-experiments/global/jetstream?project=moz-fx-data-experiments). The container installs a specific version of each library and can be uniquely identified by a SHA256 hash. A timestamp indicating when the container was uploaded is also available.
+
+Every time Jetstream runs and writes computed results to BigQuery, it tags the result tables with a last updated timestamp. 
+
+The last updated timestamp of the enrollments table can be used to determine which Docker container was the most recently published one at that time. This container can then be used for all subsequent analyses.
+
+GCR offers an [API (but no client library)](https://cloud.google.com/container-registry/docs/reference/docker-api) to access image information. Whenever jetstream starts running, it would pull in information of existing containers via this API. For every experiment that it is supposed to analyse it would determine the container hash based on the last updated timestamp of the experiments enrollment table.
+
+The hash will then be passed to the Argo workflow config, which references docker containers used for execution. The workflow config would change to something like:
+
+```yaml
+# [...]
+
+- name: analyse-experiment
+    inputs:
+      parameters:
+      - name: date
+      - name: slug
+      - name: image
+    container:
+      image: gcr.io/moz-fx-data-experiments/jetstream@{{input.parameters.image}}
+      command: [
+        jetstream, --log_to_bigquery, run,
+        "--date={{inputs.parameters.date}}",
+        "--experiment_slug={{inputs.parameters.slug}}",
+        "--dataset_id={{workflow.parameters.dataset_id}}", 
+        "--project_id={{workflow.parameters.project_id}}",
+        "--bucket={{workflow.parameters.bucket}}", 
+        "--analysis_periods={{workflow.parameters.analysis_periods_day}}",
+        "--analysis_periods={{workflow.parameters.analysis_periods_week}}",
+        "--analysis_periods={{workflow.parameters.analysis_periods_days28}}",
+        "--analysis_periods={{workflow.parameters.analysis_periods_overall}}"
+      ]
+# [...]
+```
+
+The `image` parameter corresponds to the hash. Parameterized image names are supported in Argo based on [this example](https://github.com/argoproj/argo-workflows/blob/a5581f83abd4b6d45b1bad6c9a5d471077e8427f/docs/walk-through/loops.md?plain=1#L80).
+
+Instead of using the hash, it would also be possible to start tagging docker images on deploy and use the tags instead.
```

Comment:
The only downside to doing this that I can think of is that we would need to be more diligent about updating the version number. Adding a second tag to a docker image would be very simple though.

Reply:
We could automatically version it based on the commit hash

Reply:
True, I was thinking it would make sense to keep the tags in sync with the version, but certainly not necessary.